spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/core/BigQueryTemplate.java

```diff
@@ -0,0 +1,160 @@
+ */
+public class BigQueryTemplate implements BigQueryOperations {
+
+	private final BigQuery bigQuery;
+
+	private final String datasetName;
+
+	private final TaskScheduler taskScheduler;
+
+	private boolean autoDetectSchema = true;
+
+	private WriteDisposition writeDisposition = WriteDisposition.WRITE_APPEND;
+
+	private Duration jobPollInterval = Duration.ofSeconds(2);
+
+	/**
+	 * Creates the {@link BigQuery} template..
+	 *
+	 * @param bigQuery the underlying client object used to interface with BigQuery
+	 * @param datasetName the name of the dataset in which all operations will take place
+	 * @param taskScheduler the {@link TaskScheduler} used to poll for the status of
+	 *     long-running BigQuery operations
+	 */
+	public BigQueryTemplate(BigQuery bigQuery, String datasetName, TaskScheduler taskScheduler) {
+		this.bigQuery = bigQuery;
+		this.datasetName = datasetName;
+		this.taskScheduler = taskScheduler;
+	}
+
+	/**
+	 * Sets whether BigQuery should attempt to autodetect the schema of the data when loading
+	 * data into an empty table for the first time. If set to false, the schema must be
+	 * defined explicitly for the table before load.
+	 * @param autoDetectSchema whether data schema should be autodetected from the structure
+	 *     of the data. Default is true.
+	 */
+	public void setAutoDetectSchema(boolean autoDetectSchema) {
+		this.autoDetectSchema = autoDetectSchema;
+	}
+
+	/**
+	 * Sets the {@link WriteDisposition} which specifies how data should be inserted into
+	 * BigQuery tables.
+	 * @param writeDisposition whether to append to or truncate (overwrite) data in the
+	 *     BigQuery table. Default is {@code WriteDisposition.WRITE_APPEND} to append data to
+	 *     a table.
+	 */
+	public void setWriteDisposition(WriteDisposition writeDisposition) {
+		this.writeDisposition = writeDisposition;
+	}
+
+	/**
+	 * Sets the {@link Duration} amount of time to wait between successive polls on the status
+	 * of a BigQuery job.
+	 * @param jobPollInterval the {@link Duration} poll interval for BigQuery job status
+	 *     polling
+	 */
+	public void setJobPollInterval(Duration jobPollInterval) {
+		this.jobPollInterval = jobPollInterval;
+	}
+
+	@Override
+	public ListenableFuture<Job> writeDataToTable(
+			String tableName, InputStream inputStream, FormatOptions dataFormatOptions)
+			throws IOException {
+		TableId tableId = TableId.of(datasetName, tableName);
+
+		WriteChannelConfiguration writeChannelConfiguration = WriteChannelConfiguration
+				.newBuilder(tableId)
+				.setFormatOptions(dataFormatOptions)
+				.setAutodetect(this.autoDetectSchema)
+				.setWriteDisposition(this.writeDisposition)
+				.build();
+
+		TableDataWriteChannel writer = bigQuery.writer(writeChannelConfiguration);
+
+		try (OutputStream sink = Channels.newOutputStream(writer)) {
+			// Write data from data input file to BigQuery
+			StreamUtils.copy(inputStream, sink);
+		}
+
+		if (writer.getJob() == null) {
+			throw new RuntimeException(
```

Comment:
I think we need to introduce `BigQueryException`

Reply:
:+1 

Since we talk here about some data interaction, it would be grate to consider to make this exception (or several?) as an extension of the `DataAccessException`.
See some its extensions throughout the Spring modules.

Reply:
Done, added exception.